cli/azd/internal/telemetry/storage.go

```diff
@@ -188,23 +189,29 @@ func removeIfExists(filename string) error {
 }
 
 // Scans the storage directory for any obsoleted items or temp files.
-func (stg *StorageQueue) Cleanup() {
+func (stg *StorageQueue) Cleanup(ctx context.Context, done chan (struct{})) {
```

Comment:
It feels like there might be a better way to structure this concurrency.  I wonder if using https://github.com/Azure/azure-dev/blob/dd101e45e6659dcc14c8baecad2c123e984ca611/cli/azd/pkg/async/task.go could let us remove this done parameter, and then in the one place we are calling this today with the `go` keyword we use to create a task that runs the prompt version of it and then we can just `.Await()` the task once we need the value.

Love that we now that a context and check for cancellation.

Reply:
@ellismg I think this probably warrants a team discussion on concurrency design. I was sticking to strictly `go` channels as a newcomer since it seems to be the `golang` way of implementing concurrency.

On the surface, I'm familiar that atomic + locked signal messaging is one of the most primitive types of concurrency models that can implement all kinds of patterns. I'm not sure if the lightweight async model we've implemented will have all the features we want, and whether we end up having to continually invest it. 

For example, in this PR, we wouldn't be able to implement the current logic of:
1. Run Cleanup async
2. Run a 5 second timer async
3. Wait for 5 second timer or Cleanup
4. Exit

tagging @vhvb1989 who is familiar with a lot of low-level programming :)

Reply:
We had a good discussion about this in the team meeting today (and apologies for not seeing this before that discussion).  To be clear, I don't think we need to go ahead and grab the Task pattern and use it everywhere.  My goal was more to see if there was a way to structure this without bleeding the concurrency concern into the API of `Cleanup` itself.

But I get your point about not wanting to write `Task.WhenAny` now as a reason to not jump on this.

Totally happy to stick with the send on the channel so we can use `select` with a timeout, perhaps we could move the send on the channel into the same place we invoke the go routine and do the send after `Cleanup` returns.

Reply:
> My goal was more to see if there was a way to structure this without bleeding the concurrency concern into the API of `Cleanup` itself.

We should take it offline but would love to have a complete understanding of trade-offs. I am very interested in some of the differences.

> Totally happy to stick with the send on the channel so we can use `select` with a timeout, perhaps we could move the send on the channel into the same place we invoke the go routine and do the send after `Cleanup` returns.

That's an interesting suggestion and is totally doable. I can think of some tradeoffs, and where responsibility should lie from a design perspective. I'm not completely enthusiastic about it, so I'm going to defer on this. Let's continue the discussion on `async` vs CSP, and I'll make sure to change this in a later PR if we end up deciding one way or another.